---
title: "Phylogenetic Tree Construction with FastTree2"
author: "Chloe J McGovern"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, 
                      fig.align = "center", 
                      # write figures to the figures folder
                      fig.path = "../figures/04_PhylogeneticTrees/")
```

# Inputs
1. `data/03_PreProcessing/raw_preprocessed_physeq`

# Outputs
1. `data/04_PhylogeneticTree/MAAFT_aligned_ASV.fasta`
2. `data/04_PhylogeneticTree/ASVs_unrooted.tree`

# FastTree

For amplicon-based microbial community studies, FastTree2 is widely used because it strikes a balance between speed and phylogenetic accuracy, making it ideal for datasets with thousands of sequences. For whole-genome alignments or studies requiring branch support values, more computationally intensive tools like RAxML or IQ-TREE are preferred.

Here we will use FastTree2! 


# Setup 

## Timing of this script

Let's record how long this file took to run on the class server, which we will record at the end of the script. 

```{r rmd-start}
# What time did we start running this script? 
start_time <- Sys.time()
```

## Set Seed 
```{r set-seed}
set.seed(238438)
```

## Load Packages 
```{r load-packages}
# Install phytools
#remotes::install_github("liamrevell/phytools")

# Install ggtree
#devtools::install_github("GuangchuangYu/ggtree")

# Now load them into the session 
pacman::p_load(phytools, ggtree, tidyverse, install = FALSE)
```

## 1. Load Data 

```{r load-physeq}
# Load physeq 
load("data/03_PreProcessing/raw_preprocessed_physeq.RData")
raw_preprocessed_physeq
```

# 2. Write ASV Fasta 

```{r asv-fasta}
# Pull out ASV sequences and ASV names
asv_seq_df <- 
  raw_preprocessed_physeq@tax_table %>% 
  data.frame() %>% 
  dplyr::select(ASV, ASVseqs)

#View(asv_seq_df)

# Add the ">" to make fasta header
asv_seq_df$ASV <- paste0(">", asv_seq_df$ASV)

#View(asv_seq_df)

# Create fasta object
asv_seq_fasta <- c(rbind(asv_seq_df$ASV, asv_seq_df$ASVseqs))
head(asv_seq_fasta)

# Write to a fasta file 
write(asv_seq_fasta, 
      file = "data/04_PhylogeneticTree/preprocessed_ASVs.fasta")

```

# 3. Run Alignment with MAAFT

## What is multiple sequence alignment?

Multiple sequence alignment (MSA) is the process of aligning three or more biological sequences (DNA, RNA, or protein) to identify regions of similarity. These conserved regions often indicate evolutionary relationships, structural or functional roles, and are critical for downstream analyses like phylogenetic tree building, motif discovery, or variant calling.

NOTE: **Windows:** `control` + `alt` + `enter`

```{r run-mafft, engine = 'bash', engine.opts = '-l'}
# Write bash code to run MAFFT

# Add MAFFT to PATH
export PATH=/programs/mafft/bin:$PATH

# Check working directory
echo "The working directory is $PWD"

# Set a random seed
RANDOM=238428

# Make sure output directory exists
mkdir -p data/04_PhylogeneticTree

# Run MAFFT alignment
/programs/mafft/bin/mafft --auto data/04_PhylogeneticTree/preprocessed_ASVs.fasta > data/04_PhylogeneticTree/MAFFT_aligned_ASVs.fasta

```

# 4. Calculate the phylogenetic tree with FastTree2

FastTree2 is a widely used phylogenetic tree-building program that’s particularly well-suited for microbial community analyses, especially when you’re dealing with large datasets like 16S rRNA gene sequences from amplicon surveys. 

```{r run-FastTree, engine = 'bash', engine.opts = '-l'}
# Where am I? 
echo "The working directory is $PWD"

# Load fasttree 
# Provide export path to fasttree 
export PATH=/programs/FastTree-2.1.11:$PATH

# Run Fasttree to generate phylogenetic tree 
# parameters: 
    # -nt = indicates it's a nucleotide alignment
    # -gtr = generalized time reversible substitution model 
    # -fastest speed up the model, reduce memory usage (recommended for datasets that have >50,000)
    # -log = output a log file 
    # input alignment file 
    # specify the output tree file 
FastTree -nt -gtr -fastest -log data/04_PhylogeneticTree/FastTree.log data/04_PhylogeneticTree/MAFFT_aligned_ASVs.fasta > data/04_PhylogeneticTree/ASVs_unrooted.tree
```

# Final info for Reproducibility 

## Check Render Time
```{r stop-time}
# Take the time now that we are at the end of the script
end_time <- Sys.time()
end_time 

# Echo the elapsed time
elapsed_time <- round((end_time - start_time), 3)
elapsed_time
```

## Session Information

```{r session-info}
# Ensure reproducibility with package version information
devtools::session_info()
```
