---
title: "Pruning & Rooting Phylogenetic Trees with ggTree"
author: "Chloe J McGovern"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, 
                      fig.align = "center", 
                      # write figures to the figures folder
                      fig.path = "../figures/04_PhylogeneticTree/",
                      warning = FALSE)
```

## Inputs 

1. Our previously pre-processed phyloseq object that has chloroplasts, mitochondria, reverse complements, and negative controls removed, which is located in `data/03_PreProcessing/raw_preprocessed_physeq`.
2. The raw, unrooted maximum liklihood tree that we generated using  [FastTree2](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0009490), which is located in `data/04_PhylogeneticTree/ASVs_unrooted.tree`

## Outputs 

1. The output of this file will be a S4 phloseq objected called `phytree_preprocessed_physeq`, which we will save to `data/04_PhylogeneticTree/`. 

## Timing of this script

Let's record how long this file took to run on the class server, which we will record at the end of the script. 

```{r rmd-start}
# What time did we start running this script? 
start_time <- Sys.time()
```

## Set my seed 
```{r set-seed}
# Any number can be chosen 
set.seed(238428)
```

## Load Packages 
```{r load-packages}
# install.packages("phylobase")

pacman::p_load(tidyverse, phyloseq, ggtree, phytools, tidytree, phylobase,
               install = FALSE)
```

## 1. Load Data files 
```{r load-data}
# Preprocessed phyloseq object 
load("data/03_PreProcessing/raw_preprocessed_physeq.RData")

# Inspect the phyloseq object 
raw_preprocessed_physeq

# Load in the tree! 
unrooted_tree <- read.tree("data/04_PhylogeneticTree/ASVs_unrooted.tree")

# Take a look at the tree
unrooted_tree
str(unrooted_tree)
```

# 2. Merge Unrooted Tree & `raw_preprocessed_physeq`

Now that we've loaded in both our `raw_preprocessed_physeq` and the unrooted tree, let's merge them into one phyloseq object that can be used to plot the tree with ggtree. 

```{r merge-physeq}
# Intuition check 
stopifnot(ntaxa(raw_preprocessed_physeq) == ntaxa(unrooted_tree)) # No news is good news! 

# Merge the tree with the phyloseq object 
unrooted_physeq <- 
  merge_phyloseq(raw_preprocessed_physeq, unrooted_tree)

# Let's take a look! 
unrooted_physeq
```

# 3. Visualize and inspect tree with [ggtree](https://besjournals.onlinelibrary.wiley.com/doi/full/10.1111/2041-210X.12628).

Now, let's take a look 

## Kingdom Level Tree

```{r plot-tree-unrooted, fig.width=4.5, fig.height=7}
# Make a basic tree with the domains 
kingdom_tree <- 
  ggtree(unrooted_physeq) + 
  # color tips by kingdom
  geom_tippoint(mapping = aes(color = Kingdom)) + 
  # add a title
  labs(title = "Unrooted Tree") + 
  scale_color_manual(values = c("firebrick3", "cornflowerblue", "grey")) + 
  # Move the legend to the bottom of the tree 
  theme(legend.position = "bottom")

# Look at it 
kingdom_tree

# Check if tree is rooted
is.rooted(unrooted_tree) # should be TRUE
```

The tree is unrooted! While the 515F/806R V4 primers CAN detect archaea, there are no archaea in this dataset. 

# 4. Long Branches

Are there any suspect branches in the raw, unrooted tree? Or does your tree look “clean”? Why or why not?

There is one oddly long branch that I think we should investigate. It sticks out quite further than any other branch and it's shared root node is very far away from it.  

## 4a. Evaluate Long Branches

### Kingdom Tree with Node Labels 

Let's put labels on our tree so we have a bit more flexibility

```{r kingdom-node-tree, fig.width=4.5, fig.height=7}
kingdom_node_tree <- 
  kingdom_tree + 
  # Add the node label 
  geom_text(aes(label=node), hjust= -0.5, vjust = -0.3, size = 2)

# Plot the tree
kingdom_node_tree
```

### Evaluate Node 66

```{r evaluate-long-branches-66}
# View a specific clade
# Zoom in on node that looks potentially suspicious
viewClade(kingdom_node_tree + 
          labs(title = "Kingdom Tree: Node 66"), 
          node = 66)

# This plot is too zoomed in! Let's try to zoom out by looking at ancestors
# Lets see its ancestors, which will tell us each node as it steps up. 
tidytree::ancestor(unrooted_tree, 66) #162, 160, 113, 111 ancestors

# Let's evaluate 162!
viewClade(kingdom_node_tree, 162) #162 related to 65 and long branch 66
```


```{r pull-node66}
# Or if there are two nodes that we can see we can find its MRCA 
## create tibble of tree to identify offspring
tree_tib <- as_tibble(unrooted_physeq@phy_tree)

# Inspect
head(tree_tib) #node 66 related to 162 parent, ASV_0093
str(tree_tib)

# lets look at the long branch
mrca_node_66 <- tidytree::MRCA(unrooted_tree, .node1 = 66, .node2 = 162)

# Inspect
mrca_node_66

## create dataframe to view ASVs and taxa info and to pull the ASVs
node_162 <- 
  offspring(tree_tib, mrca_node_66, tiponly = TRUE) %>% 
  as.data.frame() %>% 
  pull(label)

# Inspect
node_162

# Now, let's pull out the taxonomy information, including the ASV sequence!
node_162_df <- 
  unrooted_physeq %>%
  tax_table() %>%
  as.data.frame() %>% 
  filter(ASV %in% node_162) 

# Take a look: 
glimpse(node_162_df)

# look at counts of the ASVs
unrooted_physeq %>%
  subset_taxa(., ASV %in% node_162_df$ASV) %>%
  otu_table() %>%
  data.frame() %>%
  colSums()

# Lets BLAST our sequences to investigate if they are real or not
node_162_df 
```

Node 66 is attached to ASV_0093. The sequence associated is "TACGTAGGGGGCGAGCGTTGTCCGGAGTTACTGGGCGTAAAGGGCCTGCAGGCGGTCTGGTGCGTTTCGGCGGACAGCTGGTGGCTTAACTACCAGAGGGGCCGAAAGACGATCAGACTTGAGGGCCAGAGAGGGACCTGGAATTCCGGGTGGAGCGGTGAAATGCGTAGAGATCCGGAGGAACACCGAAGGCGAAGGCAGGGTCCTGGCTGGCGACTGACGCTGAGGGGCGAAAGCTAGGGGAGCGAACGGG" 

Let's BLAST it confirm that this is a legitimate ASV in this system.  

# BLAST

Now, copy and paste the ASV sequences into NCBI’s Standard Nucleotide BLAST. I had re-blasted the sequence for ASV 93 and it turns out that it is NOT a known sequence. However, it has a 100% query cover percentage to tens of sequences noted as "unidentified bacterium", as well as a very small E value, indicating this ASV is significant and unlikely to be due to chance. 

## 4b. Prune your tree

Were there any long branches in the unrooted tree? If so, did you decide to remove any ASVs from your tree? Why or why not? What data convinced you to remove each of the ASVs?

Branch 66 looks to be a long but after inspecting the ASV and BLAST-ing the sequence, I have decided to keep it. So I will fast forward to midpoint rooting analysis since I have no archaea. 

# 5. Root the Tree

## Why do we root trees in microbial community analyses?

In microbial ecology, rooted trees are especially useful for:

1. **Calculating Phylogenetic Diversity Metrics**: Metrics like Faith’s Phylogenetic Diversity and the UniFrac distances require a rooted tree to calculate how much evolutionary history is shared between communities. Without a root, the distances between taxa don’t make biological sense in terms of lineage or shared ancestry.
2. **Comparing Community Structure:** Rooting allows you to identify lineages that dominate or are lost between samples, trace evolutionary shifts in microbial composition, interpret ecological or functional patterns across a gradient (*e.g.,* salinity, pH, antibiotic exposure). 
3. **Biological Interpretation** Rooted trees are required for the following interpretations, which microbial community ecologists often want to know: 
    - *Are these microbes derived from a common ancestor?*
    - *Do the samples diverge from a known lineage (e.g., Proteobacteria)*?
    - *What’s the evolutionary placement of unknown organisms?*
    
Because we only have bacteria at the kingdom level, we will be midpoint rooting this tree. 
    
**Midpoint rooting:** Places the root at the midpoint of the longest path in the tree (useful when no clear outgroup exists). 
    - *Pros:*
        - Simple and automatic — no need for prior biological knowledge.
        - Works well for diverse microbial datasets where outgroup info is unclear or unavailable.
        - **Best for**: Exploratory analyses, diversity metrics (like UniFrac), or when working with just bacteria/just archaea and you don’t have a good outgroup.
    - *Cons:*
        - Assumes a molecular clock — that all lineages evolve at similar rates.
        - Can misplace the root if there’s rate variation across branches.

## 5a. Midpoint Root 

**Before midrooting the tree:**

1. Investigate and pull the ASVs in your long branches and BLAST to confirm if these are real bacterial ASVs, as mentioned above in the "long branches" section.

2. Remove any unwanted ASVs using the prune function. **Long nodes (branches) are outliers**! If we keep them in the tree, bias will be introduced into the tree reconstruction creating the **long branch attraction** error where distantly related taxa appear closely related due to a high rate of evolution on long branches. 

### 6b. Merge phyloseq: Midpoint Rooted Tree

```{r midpoint-root-tree-physeq}
# Check to see if tree is rooted.. expect it to be false
is.rooted(phy_tree(unrooted_physeq))

# Let's midpoint root the tree
midpoint_rooted_tree <- 
  unrooted_physeq %>%
  phy_tree() %>%
  phytools::midpoint.root()

# Is the new tree rooted? Should be TRUE!
is.rooted(midpoint_rooted_tree)

# Assign to a new phyloseq object!
# Merge tree with the raw_preprocessed physeq
midpoint_rooted_physeq <-
  merge_phyloseq(raw_preprocessed_physeq, midpoint_rooted_tree)
```

Now, we have a merged phyloseq object!  

### Plot Midpoint Rooted Tree

```{r plot-midpoint-root-tree,fig.width=4.5, fig.height=7}
# Quick inspection of tree 
midpoint_rooted_tree_plot <-
  ggtree(midpoint_rooted_physeq) + 
  geom_tippoint(mapping = aes(color = Kingdom)) + 
  geom_tippoint(aes(color = Kingdom)) + 
  labs(title = "Midpoint Rooted Tree") + 
  scale_color_manual(values = c("firebrick3", "cornflowerblue", "grey"))

# Look at the tree 
midpoint_rooted_tree_plot

# Add nodes 
midpoint_rooted_tree_node_plot <- 
  midpoint_rooted_tree_plot + 
  geom_text(aes(label = node), hjust = -0.5, vjust = -0.3, size = 1.5) 

# View it 
midpoint_rooted_tree_node_plot
```

Now that you’ve rooted your tree, does the tree seem “finalized”? Are there any other branches/ASVs to inspect? Why or why not? 

The tree does seem finalized. I do worry about the weighting of branch length in downstream analysis if I should remove node 66 or trust my gut and the BLAST that tells me I should keep it for now since it is related to the potential community of (poorly processed) cheese. Other than that one, I think the rest fall within reason of their respective branches. 

## 7. Save phyloseq objects

### Save multiple phyloseq objects as a list

This will keep two phyloseq objects in on .RData object that will allow us to easily load in both trees into an R session at once. 

```{r save-phyloseq-objects-list}
# Take a quick look
midpoint_rooted_physeq

# Save both phyloseq objects to one .RData file 
save(midpoint_rooted_physeq, 
     file = "data/04_PhylogeneticTree/midpoint_rooted_physeq.RData")
```

Now that we have our rooted trees inside of a phyloseq object, we can go forth and begin our next analysis!

# Final info for Reproducibility 

## Check Render Time
```{r stop-time}
# Take the time now that we are at the end of the script
end_time <- Sys.time()
end_time 

# Echo the elapsed time
elapsed_time <- round((end_time - start_time), 3)
elapsed_time
```

## Session Information

```{r session-info}
# Ensure reproducibility with package version information
devtools::session_info()
```


